### ISSUE TITLE
question on performance bottleneck

### ISSUE THREAD

**Issue by Chen-Cai-OSU** [e3nn/e3nn#165]
Hello,

I am interested in understanding the bottleneck of performance for e3nn. I did some profile and here is my finding:

1) during training, ```loss.backward()``` takes roughly 2x time of the forward.
2) for the forward, it seems that most time is spent on the TensorProduct module.

I am profiling example/point/tetris_torch_geo.py. I also tired profiling my code for different data, 1) also hold.

If 1) and 2) are representative, then It seems that at least for training stage of e3nn, the bottleneck is not the tensorproduct (I am sure this part can be further improved), but the backward gradient calculation. It seems that right now it is handled by the PyTorch autograd. Do you think there is a big room for improvement for ```loss.backward```? (from https://github.com/e3nn/e3nn/issues/28 it looks like customer backward is faster?)

Thank you very much and happy new year!

**Comment by mariogeiger** [comment]
Hello,

It look natural that the backward takes 2x the forward. 

If you multiply two matrices you have at the forward `C=AB` and at the backward you have `dA=dC Bt` and `dB=At dC` which gives twice amount of computation for the backward. 

I also observed that the bottleneck is the tensor product. The code is now different to what @bkmi benchmarked. Custom backward might improve the performance but I doubt. Custom cuda kernels should be useful and @L-sky is working on it.

**Comment by Chen-Cai-OSU** [comment]
Hello,

Thank you very much for the explanation!
Another question: if custom cuda kernel can improve the forward process, will it also improve the backward process?

**Comment by mariogeiger** [comment]
Yes, if we implement the forward pass with cuda we will need to implement the backward pass as well.

**Comment by mariogeiger** [comment]
![image](https://user-images.githubusercontent.com/333780/103444870-7bf91600-4c6d-11eb-8207-f5e8557ca59f.png)

You can use [the pytorch profiler](https://pytorch.org/tutorials/recipes/recipes/profiler.html) to have a better look at what takes time to execute
